From 605dfc3fa09b280ab8ac44337cdf7f6d97345e5a Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Thu, 21 Sep 2023 12:31:56 +0200
Subject: [PATCH 0111/1010] genirq: irq_pipeline: fix inconsistent lockdep
 check

Acquiring hard locks from inband entails disabling hard irqs without
requiring to stall the stage.

Fix false positives w/ DEBUG_HARD_LOCKS on in check_flags() according
to this case.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 include/linux/irqstage.h | 7 ++++---
 kernel/irq/pipeline.c    | 7 ++++---
 kernel/locking/lockdep.c | 6 ++++--
 3 files changed, 12 insertions(+), 8 deletions(-)

diff --git a/include/linux/irqstage.h b/include/linux/irqstage.h
index a73b91e909..5fb6b7037a 100644
--- a/include/linux/irqstage.h
+++ b/include/linux/irqstage.h
@@ -318,7 +318,7 @@ static __always_inline void oob_irq_restore(unsigned long x)
 
 bool stage_disabled(void);
 
-bool stage_disabled_flags(unsigned long irqstate);
+bool stage_disabled_flags(unsigned long irqstate, bool *stalled);
 
 unsigned long test_and_lock_stage(int *irqsoff);
 
@@ -367,9 +367,10 @@ static __always_inline bool stage_disabled(void)
 	return irqs_disabled();
 }
 
-static __always_inline bool stage_disabled_flags(unsigned long flags)
+static __always_inline bool stage_disabled_flags(unsigned long flags, bool *stalled)
 {
-	return irqs_disabled_flags(flags);
+	*stalled = irqs_disabled_flags(flags);
+	return *stalled;
 }
 
 static __always_inline void irq_post_inband(unsigned int irq)
diff --git a/kernel/irq/pipeline.c b/kernel/irq/pipeline.c
index f8d0e8794c..89119db8da 100644
--- a/kernel/irq/pipeline.c
+++ b/kernel/irq/pipeline.c
@@ -490,12 +490,13 @@ EXPORT_SYMBOL_GPL(stage_disabled);
  *      though, since this is guaranteed by the implementation of
  *      test_and_lock_stage().
  */
-noinstr bool stage_disabled_flags(unsigned long irqstate)
+noinstr bool stage_disabled_flags(unsigned long irqstate, bool *stalled)
 {
 	unsigned long flags;
-	int stalled;
+	int _stalled;
 
-	flags = irqs_split_flags(irqstate, &stalled);
+	flags = irqs_split_flags(irqstate, &_stalled);
+	*stalled = _stalled;
 	return hard_irqs_disabled_flags(flags) || stalled;
 }
 EXPORT_SYMBOL_GPL(stage_disabled_flags);
diff --git a/kernel/locking/lockdep.c b/kernel/locking/lockdep.c
index 2a360f812f..1604c09eef 100644
--- a/kernel/locking/lockdep.c
+++ b/kernel/locking/lockdep.c
@@ -5568,6 +5568,8 @@ static void __lock_unpin_lock(struct lockdep_map *lock, struct pin_cookie cookie
 static noinstr void check_flags(unsigned long flags)
 {
 #if defined(CONFIG_PROVE_LOCKING) && defined(CONFIG_DEBUG_LOCKDEP)
+	bool stalled;
+
 	/*
 	 * irq_pipeline: we can't and don't want to check the
 	 * consistency of the irq tracer when running the interrupt
@@ -5580,8 +5582,8 @@ static noinstr void check_flags(unsigned long flags)
 	/* Get the warning out..  */
 	instrumentation_begin();
 
-	if (stage_disabled_flags(flags)) {
-		if (DEBUG_LOCKS_WARN_ON(lockdep_hardirqs_enabled())) {
+	if (stage_disabled_flags(flags, &stalled)) {
+		if (DEBUG_LOCKS_WARN_ON(lockdep_hardirqs_enabled() && stalled)) {
 			printk("possible reason: unannotated irqs-off.\n");
 		}
 	} else {
-- 
2.30.2

